pkg/querier/frontend2/frontend.proto

```diff
@@ -0,0 +1,94 @@
+syntax = "proto3";
+
+package frontend2;
+
+option go_package = "frontend2";
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+import "github.com/weaveworks/common/httpgrpc/httpgrpc.proto";
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// Scheduler interface exposed to Queriers. (Previously called Frontend)
+service SchedulerForQuerier {
+    // After calling this method, both Querier and Scheduler enters a loop, in which querier waits for
+    // a "SchedulerToQuerier" messages containing HTTP requests and processes them. After processing the request,
+    // querier signals that it is ready to accept another one by sending empty QuerierToScheduler message.
+    //
+    // Long-running loop is used to detect broken connection between scheduler and querier. This is important
+    // for scheduler to keep a list of connected queriers up-to-date.
+    rpc QuerierLoop(stream QuerierToScheduler) returns (stream SchedulerToQuerier) { };
+}
+
+// Querier reports its own clientID when it connects, so that scheduler knows how many *different* queriers are connected.
+// To signal that querier is ready to accept another request, querier sends empty message.
+message QuerierToScheduler {
+    string querierID = 1;
+}
+
+message SchedulerToQuerier {
+    uint64 queryID = 1;
+    httpgrpc.HTTPRequest httpRequest = 2;
+
+    // Where should querier send HTTP Response to (using FrontendForQuerier interface).
+    string frontendAddress = 3;
+
+    // User who initiated the request. Needed to send reply back to frontend.
+    string userID = 4;
+}
+
+// Frontend interface exposed to Queriers. Used by queriers to report back the result of the query.
+service FrontendForQuerier {
+    rpc QueryResult (QueryResultRequest) returns (QueryResultResponse) { };
+}
+
+message QueryResultRequest {
```

Comment:
In Cortex, userID in gRPC is mostly transferred via context (and call metadata). That isn't possible for long-running calls like `FrontendLoop` and `QuerierLoop`, because they are used for multiple users.

In case of `QueryResult` call however we *do* transfer userID in the context again. That's why it is not visible in the `QueryResultRequest`. I will add comment about it.